### ISSUE TITLE
Error when non-admin user accesses admin console

### ISSUE THREAD

**Issue by HannahBrown** [keycloak/keycloak#31638]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

admin/api

### Describe the bug

My installation has multiple users who have access to the admin console for specific realms, not including the master realm. For example, a user has permissions to manage other users in a different realm. Previously when that user would log in, they would view the Keycloak homepage and be able to select the necessary page in a different realm. Now, those users are able to successfully log in still, but receive an error on the homepage. While they are able to navigate to other pages (including ones that should work), the error doesn't go away.

### Version

25.0.1

### Regression

- [X] The issue is a regression

### Expected behavior

Users who have admin access to specific realms should be able to access those realms in the admin ui without error.

### Actual behavior

Users who do not have the master realm's `view-realm` role receive an error when logging in, and the refresh button in the error does not help. Additionally, navigating to other pages they should have access to still gives the same error.

### How to Reproduce?

1. Create a new user in the master realm.
2. Give that user non-master realm permissions without any master realm permissions.
3. Log in to the admin console as that user.

### Anything else?

Previously, the master realm users were able to see Keycloak statistics, such as version and memory allocation. I agree that users who do not have access to the master realm should not be able to see that information, but they should not get an error just by logging in.

**Comment by jonkoops** [comment]
Thanks for reporting this issue, we certainly don't want to introduce regressions like this.

> Now, those users are able to successfully log in still, but receive an error on the homepage. 

What is the error you are seeing? Are there any server or browser (console or network) errors? Did you try the latest version of Keycloak (25.0.2)?

> Give that user non-master realm permissions without any master realm permissions.

I am sorry, but this sentence is kinda hard to parse. Could you provide more detailed step by step reproduction details?

**Comment by HannahBrown** [comment]
I'm sorry for the lack of clarity.

> What is the error you are seeing? Are there any server or browser (console or network) errors? Did you try the latest version of Keycloak (25.0.2)?

I did try 25.0.2, and I still got the same error. On first login, the screen looks like this:
![master](https://github.com/user-attachments/assets/daf09258-1ece-4e1d-9897-655914f99b32)
It is still possible to use the sidebar, so I tried to go to a realm that the user has access to. It looks like this:
![bonus](https://github.com/user-attachments/assets/6e5376db-bdfa-48c2-9944-fba9200196b4)

> I am sorry, but this sentence is kinda hard to parse. Could you provide more detailed step by step reproduction details?

This is the list of roles that the user has:
![roles](https://github.com/user-attachments/assets/1874f6c2-264f-4d2a-8c65-f724206d7b46)
The user is in the master realm, but does not have any permissions to view or update the master realm. Instead, they have specific admin privileges for a different realm. I found that it is possible to resolve the error by giving the user a master-realm role, as shown below:
![roles-master](https://github.com/user-attachments/assets/44bc3b88-e245-4698-977e-91198c3701a2)
However, I would prefer that this class of user did not have any master-realm permissions.

Please let me know if additional information is needed.

**Comment by sguilhen** [comment]
I can confirm the behavior has changed from Keycloak 24 - in that version, once you login with a user without privileges on master realm is greeted with the Welcome to Keycloak screen, along with Server Info and Provider Info tabs that probably shouldn't be visible. But at least the user can switch to the realm where they have permissions (for example to manage users on that realm).

In Keycloak 25, the user gets the permission error that prevents them from even switching to the realm where they have permissions, so something changed with the permission evaluation. I don't mind suppressing the Server Info and Provider Info tabs on master if the user has no permissions on that realm, but it keeping the Welcome to Keycloak screen with the possibility to switch to a realm the user has permissions seems a good idea.

Not sure yet if it is the admin console that is enforcing permissions that it wasn't enforcing before, or if something changed in the backend - needs some investigation. @edewit Are you aware of any changes in the UI side?

**Comment by edewit** [comment]
I am not, but shouldn't you login on the realm you have the permissions on instead of the master realm, this would prevent one to have to switch after login

**Comment by sguilhen** [comment]
I was thinking about that too. @HannahBrown I think it would be better to create these users in the realm where they have permissions (instead of the master realm) and then login to the console using the right realm.

**Comment by HannahBrown** [comment]
I discussed this with the person who originally set up the roles. The issue here is that we have multiple people with little technical know-how who have these roles for several realms. The roles were created in the master realm so that they did not have to go to multiple sites for the same thing. 
I'll discuss this with the various supervisors, but I'm not sure I can convince them that changing this workflow is the best solution.

**Comment by edewit** [comment]
@HannahBrown so you are saying that this used to work?

**Comment by HannahBrown** [comment]
Yes, this used to work. @sguilhen was able to confirm the change in behavior without the realm export. If you do need it, I'll try to get it during downtime tonight.

**Comment by pedroigor** [comment]
We have plans to review the fine-grained admin in order to make it supported. The issue is now in our backlog so that we can review it once we start reviewing the feature.

**Comment by simse07** [comment]
Hi, we also have the same issue.
Our setup heavily depends on the login via the master realm.
We provide an user federation in the master realm and assign the corresponding realm admin roles depending on the ldap groups in the master realm to each admin user.
This gives us the possibility to provide a single login to all of our 20+ realms, without the need of handling the admin users per realm.

As a workaround you can directly go the the realm page via the master realm:
https://..../admin/master/console/#/REALM

**Comment by pedroigor** [comment]
@edewit It seems to be UI related because if the page is refreshed (not clicking on the link but using a browser refresh) after switching to the realm then you are allowed to access the realm.

Perhaps selecting a realm should make sure the realm access is fetched again for the selected realm?

**Comment by pedroigor** [comment]
@edewit Wdyt about https://github.com/keycloak/keycloak/pull/33500?

**Comment by sguilhen** [comment]
I think the proposed solution makes sense, @pedroigor - without reloading the user gets stuck with no permssions

**Comment by HannahBrown** [comment]
The solution should work for our use case. Thank you for working on this.
